src/parser.ml

```diff
@@ -989,19 +989,58 @@ let link_label allow_balanced_brackets st =
   in
   loop 0 false
 
+(* based on https://erratique.ch/software/uucp/doc/Uucp/Case/index.html#caselesseq *)
 let normalize s =
-  let buf = Buffer.create (String.length s) in
-  let rec loop start seen_ws i =
-    if i >= String.length s then Buffer.contents buf
-    else
-      match s.[i] with
-      | ' ' | '\t' | '\010' .. '\013' -> loop start true (succ i)
-      | _ as c ->
-          if (not start) && seen_ws then Buffer.add_char buf ' ';
-          Buffer.add_char buf (Char.lowercase_ascii c);
-          loop false false (succ i)
+  let canonical_caseless_key s =
+    let b = Buffer.create (String.length s * 2) in
+    let to_nfd_and_utf_8 =
+      let n = Uunf.create `NFD in
+      let rec add v =
+        match Uunf.add n v with
+        | `Await | `End -> ()
+        | `Uchar u ->
+            Uutf.Buffer.add_utf_8 b u;
+            add `Await
+      in
+      add
+    in
+    let add =
+      let n = Uunf.create `NFD in
+      let rec add v =
+        match Uunf.add n v with
+        | `Await | `End -> ()
+        | `Uchar u ->
+            (match Uucp.Case.Fold.fold u with
+            | `Self -> to_nfd_and_utf_8 (`Uchar u)
+            | `Uchars us -> List.iter (fun u -> to_nfd_and_utf_8 (`Uchar u)) us);
+            add `Await
+      in
+      add
+    in
+    let fold () =
+      let start = ref true in
+      let seen_ws = ref false in
+      let add_uchar () _ = function
+        | `Malformed _ ->
+            start := false;
+            seen_ws := false;
+            add (`Uchar Uutf.u_rep)
+        | `Uchar u as uchar ->
+            if Uucp.White.is_white_space u then seen_ws := true
+            else (
+              if (not !start) && !seen_ws then add (`Uchar (Uchar.of_char ' '));
```

Comment:
I'm a bit unsure whether using mutable state here is producing the most readable (or performant) fold. I've a suspicion we could use the accumulator arg of the folded function to track the state of whether we've seen white space, and I'm not sure convinced that `start` is needed at all.

I'll see if I can sketch out what I have in mind later this week, unless you point out something essential I'm missing :)

Reply:
That's what I did in the first version I implemented but since the return value is not used I thought perhaps it's just fine to use mutation.

I made a version in https://github.com/ocaml/omd/pull/277/commits/4e96339463976f4fd1d7e308180a4a85c973b42d that uses a tuple of two boolean to track the `start` and `seen_ws` values. I'm not sure that's ideal since we can easily mix them, maybe a record instead is better?

Reply:
See https://github.com/ocaml/omd/pull/277/commits/e51ad122add893419765416e6876f2a54d21b27f for the version that uses record. Let me know which one you prefer

Reply:
I see what you mean. I had the false impression we could do away with `start`. Thank you for explaining the need. Given that requirement, I don't have a preference over the record vs. the mutable refs and I'm happy to go with whichever you prefer here.

Sorry for the red herring here!